src/megatron/energon/flavors/webdataset/indexing.py

```diff
@@ -158,6 +160,130 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         self.close()
 
 
+class SqliteIndexReader:
+    """Reads samples from an SQLite database created by SqliteIndexWriter.
+
+    The database contains a table with the following schema:
+    - samples(id INTEGER PRIMARY KEY AUTOINCREMENT,
+              tar_file_id INTEGER,
+              sample_key TEXT,
+              sample_index INTEGER,
+              byte_offset INTEGER,
+              byte_size INTEGER)
+    """
+
+    sqlite_path: EPath
+    db: ThreadLocalSqlite
+
+    def __init__(self, sqlite_path: EPath):
+        """Initialize the SQLite database reader.
+
+        Args:
+            sqlite_path: Path to the SQLite database file
+        """
+        self.sqlite_path = sqlite_path
+
+        # Initialize SQLite connection
+        path = str(self.sqlite_path)
+        # Only supporting local file system, because sqlite does not support remote file systems
+        assert path.startswith("/"), (
+            f"SQLite path must be absolute local file system path: {self.sqlite_path}"
+        )
+
+        self.db = ThreadLocalSqlite(path)
+
+    def get_sample_count(self) -> int:
+        """Get the total number of samples in the database."""
+        assert self.db is not None, "Database is closed"
+
+        count = self.db.select_one("SELECT COUNT(*) FROM samples")
+        return count[0] if count else 0
+
+    def get_sample_by_index(self, idx: int) -> Any:
+        """Get a sample by its numeric index.
+
+        Args:
+            idx: 0-based index of the sample
+
+        Returns:
+            Tuple of (tar_file_id, sample_key, sample_index, byte_offset, byte_size)
+        """
+        assert self.db is not None, "Database is closed"
+
+        # SQLite uses 1-based indexing for ROWID, so add 1 to idx
+        sample = self.db.select_one(
+            "SELECT tar_file_id, sample_key, sample_index, byte_offset, byte_size "
+            "FROM samples WHERE id = ?",
```

Comment:
```suggestion
            "FROM samples WHERE sample_index = ?",
```
Is this correct? Shouldn't this be `sample_index = ?`?

Reply:
Correction: The sample_index is the local index in the tar file. I fear we do not have the global index? The id is in the order of creation, but that is not guaranteed to be one tar file after the other, because we index multiple tar files at once? I think we should fix this.